bgpd/bgp_labelpool.c

```diff
@@ -0,0 +1,529 @@
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <zebra.h>
+
+#include "log.h"
+#include "memory.h"
+#include "stream.h"
+#include "mpls.h"
+#include "vty.h"
+#include "skiplist.h"
+#include "workqueue.h"
+#include "zclient.h"
+
+#include "bgpd/bgpd.h"
+#include "bgpd/bgp_labelpool.h"
+#include "bgpd/bgp_debug.h"
+
+/*
+ * Definitions and external declarations.
+ */
+extern struct zclient *zclient;
+
+/*
+ * Remember where pool data are kept
+ */
+static struct labelpool *lp;
+
+/* request this many labels at a time from zebra */
+#define LP_CHUNK_SIZE	50
+
+struct lp_cbq_item {
+	int		(*cbfunc)(mpls_label_t label, void *lblid, bool alloc);
+	int		type;
+	mpls_label_t	label;
+	void		*labelid;
+	bool		allocated;	/* false = lost */
+};
+
+static wq_item_status lp_cbq_docallback(struct work_queue *wq, void *data)
+{
+	struct lp_cbq_item *lcbq = data;
+	int rc;
+	int debug = BGP_DEBUG(vpn, VPN_LEAK_LABEL);
+
+	if (debug)
+		zlog_debug("%s: calling callback with labelid=%p label=%u allocated=%d",
+			__func__, lcbq->labelid, lcbq->label, lcbq->allocated);
+
+	if (lcbq->label == MPLS_LABEL_NONE) {
+		/* shouldn't happen */
+		zlog_err("%s: error: label==MPLS_LABEL_NONE", __func__);
+		return WQ_SUCCESS;
+	}
+
+	rc = (*(lcbq->cbfunc))(lcbq->label, lcbq->labelid, lcbq->allocated);
+
+	if (lcbq->allocated && rc) {
+		/*
+		 * Callback rejected allocation. This situation could arise
+		 * if there was a label request followed by the requestor
+		 * deciding it didn't need the assignment (e.g., config
+		 * change) while the reply to the original request (with
+		 * label) was in the work queue.
+		 */
+		if (debug)
+			zlog_debug("%s: callback rejected allocation, releasing labelid=%p label=%u",
+				__func__, lcbq->labelid, lcbq->label);
+
+		uintptr_t lbl = lcbq->label;
+		void *labelid;
+		struct lp_lcb *lcb;
+
+		/*
+		 * If the rejected label was marked inuse by this labelid,
+		 * release the label back to the pool.
+		 *
+		 * Further, if the rejected label was still assigned to
+		 * this labelid in the LCB, delete the LCB.
+		 */
+		if (!skiplist_search(lp->inuse, (void *)lbl, &labelid)) {
+			if (labelid == lcbq->labelid) {
+				if (!skiplist_search(lp->ledger, labelid,
+					(void **)&lcb)) {
+					if (lcbq->label == lcb->label)
+						skiplist_delete(lp->ledger,
+							labelid, NULL);
+				}
+				skiplist_delete(lp->inuse, (void *)lbl, NULL);
+			}
+		}
+	}
+
+	return WQ_SUCCESS;
+}
+
+static void lp_cbq_item_free(struct work_queue *wq, void *data)
+{
+	XFREE(MTYPE_BGP_LABEL_CBQ, data);
+}
+
+static void lp_lcb_free(void *goner)
+{
+	if (goner)
+		XFREE(MTYPE_BGP_LABEL_CB, goner);
+}
+
+static void lp_chunk_free(void *goner)
+{
+	if (goner)
+		XFREE(MTYPE_BGP_LABEL_CHUNK, goner);
+}
+
+void lp_init(struct thread_master *master, struct labelpool *pool)
```

Comment:
to rename : lp_init => bgp_lp_init
at least, find a way to ensure that we can not do mistake with zebra label alloactor

Reply:
I am OK with s/lp_/bgp_lp_/ but I thought someone might like to shift this code to lib so it would be available to other daemons. If nobody expresses interest in that shift, I will make the substitution.

Reply:
Moving this code to _lib/_ is probably a good idea. I found very interesting this abstraction you implemented on top of the label manager and I'll probably want to use it on ldpd. Asynchronous label allocation is a bit tricky to get right on the client daemons but overall it's the correct thing to do for a variety of reasons. Thanks for working on this.